workflow/cron/operator.go

```diff
@@ -117,58 +116,68 @@ func getWorkflowObjectReference(wf *v1alpha1.Workflow, runWf *v1alpha1.Workflow)
 }
 
 func (woc *cronWfOperationCtx) persistUpdate() {
-	_, err := woc.cronWfIf.Update(woc.cronWf)
+	if woc.origCronWf.ResourceVersion != woc.cronWf.ResourceVersion {
+		woc.log.Error("cannot update cron workflow with mismatched resource versions")
+		return
+	}
+	cronWf, err := woc.cronWfIf.Update(woc.cronWf)
 	if err != nil {
-		if errors.IsConflict(err) {
-			reapplyErr := woc.reapplyUpdate()
-			if reapplyErr != nil {
-				woc.log.WithError(reapplyErr).WithField("original error", err).Error("failed to update CronWorkflow after reapply attempt")
-			}
-		} else {
+		if !errors.IsConflict(err) {
 			woc.log.WithError(err).Error("failed to update CronWorkflow")
+			return
+		}
+		var reapplyErr error
+		cronWf, reapplyErr = woc.reapplyUpdate()
+		if err != nil {
+			woc.log.WithError(reapplyErr).WithField("original error", err).Error("failed to update CronWorkflow after reapply attempt")
+			return
 		}
 	}
+	woc.cronWf = cronWf
```

Comment:
👍

Reply:
Actually, I'm not too sure what the purpose of this is. In theory, once `persistUpdates` is called, it should be the end of the cron operation context and nothing else should access `woc.cronWf`. Could you explain a bit why this line is necessary?

Reply:
I've included this so that subsequent calls to this func will have the updated version, and as a result cannot accidentally reapplyUpdate twice.

Reply:
> I've included this so that subsequent calls to this func will have the updated version, and as a result cannot accidentally reapplyUpdate twice.

In that case wouldn't the guard rail here (https://github.com/argoproj/argo/pull/3883#discussion_r478680535) always fail? The resource version of the new `woc.cronWf` will always be different than the old.

It seems like we don't intend to call `persistUpdates` more than once at all. Should we instead ensure that it doesn't happen altogether?

Reply:
My goals to make sure we don't regress by putting in place checks. I would want the guard condition to fail - as it indicates a programming error.